### PR TITLE
[POAE7-2775] HashTable implementation of string

### PR DESCRIPTION
**Pull request by Deegue**
### What changes were proposed in this pull request?
1. Add StringHashMap and StringHashTable implementation.
2. Add String support for AggregationHashTable.
3. Add short string and long string unit tests for two suites.

### Why are the changes needed?
Improvement.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UTs.

### CODE REVIEW

File: cpp/src/cider/common/interpreters/AggregationHashTable.cpp [hunk 1]
```diff
@@ -72,48 +72,57 @@ AggregateDataPtr AggregationHashTable::get(int8_t* raw_key) {
...
       double key_v = (reinterpret_cast<double*>(key.getAddr()))[0];
       if (agg_ht_double_[key_v] == nullptr) {
         agg_ht_double_[key_v] = allocator.allocate(init_len_);
         std::memcpy(agg_ht_double_[key_v], init_val_, init_len_);
       }
       return agg_ht_double_[key_v];
+    } else if (SQLTypes::kCHAR == key_types_[i].get_type() ||
+               SQLTypes::kVARCHAR == key_types_[i].get_type() ||
+               SQLTypes::kTEXT == key_types_[i].get_type()) {
+      float key_v = (reinterpret_cast<float*>(key.getAddr()))[0];
```
Comment (jikunshang) [inline cpp/src/cider/common/interpreters/AggregationHashTable.cpp:120]: why cast to float

File: cpp/src/cider/tests/CiderNewHashTableTest.cpp [hunk 2]
```diff
@@ -521,6 +522,84 @@ TEST_F(CiderNewHashTableTest, aggUInt64LargeDatasetTest) {
...
+    CHECK_EQ(ht_long_str[keys[i]].getCount(), 1);
+  }
+
+  std::vector<std::string> keys2;
+  for (size_t i = 0; i < 5; i++) {
+    std::string tmp_str;
+    for (int j = 0; j < i * 100; j++) {
+      tmp_str = tmp_str + base_str;
+    }
+    keys2.push_back(tmp_str);
```
Comment (jikunshang) [inline cpp/src/cider/tests/CiderNewHashTableTest.cpp:588]: keys2.push_back(keys[i]);

File: cpp/src/cider/tests/CiderNewHashTableTest.cpp [hunk 3]
```diff
@@ -521,6 +522,84 @@ TEST_F(CiderNewHashTableTest, aggUInt64LargeDatasetTest) {
...
+  }
+  for (int i = 0; i < keys.size(); i++) {
+    CHECK_EQ(ht_short_str[keys[i]].getSum(), 60);
+    CHECK_EQ(ht_short_str[keys[i]].getCount(), 2);
+    CHECK_EQ(ht_short_str[keys[i]].getAvg(), 30);
+  }
+}
+
+TEST_F(CiderNewHashTableTest, aggLongStringTest) {
+  using AggregatedHashTableForLongStringKey = HashMapWithSavedHash<StringRef, Block>;
```
Comment (jikunshang) [inline cpp/src/cider/tests/CiderNewHashTableTest.cpp:556]: I think StrringRef usage is not safe. https://github.com/intel/BDTK/blob/main/cpp/src/cider/common/base/StringRef.h#L54

```
AggregatedHashTableForLongStringKey map;
{ 
std::string s = "a long string";
auto x = map[s];  // implicit conversion from std::string to StringRef, will record pointer of std::string
} // s deconstruct
auto x =map[s] // not safe!
```

Comment (Deegue) [reply cpp/src/cider/tests/CiderNewHashTableTest.cpp:556]: Yes, but `StringRef` is some kind a wrapper of `std::string`, which should have the same lifecycle. Besides, `std::string` is only used for test currently.

File: cpp/src/cider/tests/CiderNewHashTableTest.cpp (hunk 2 above)
Comment (jikunshang) [reply cpp/src/cider/tests/CiderNewHashTableTest.cpp:588]: push_back will append a copy of `tmp_str`. Maybe you want this. `keys2.push_back(std::move(tmp_str));`
https://en.cppreference.com/w/cpp/container/vector/push_back

File: cpp/src/cider/tests/CiderNewHashTableTest.cpp (hunk 3 above)
Comment (jikunshang) [reply cpp/src/cider/tests/CiderNewHashTableTest.cpp:556]: My point is this API/usage is risky. you need to tell your user this limitation.

**Review by jikunshang — APPROVED** [review]

**Review by harborn — APPROVED** [review]

**Pull request merged**